File: src/normalize.cpp
#include "normalize.h"

#include <algorithm>
#include <array>
#include <cctype>

namespace malapi {
namespace {

bool is_lower(char c) {
    return c >= 'a' && c <= 'z';
}

bool is_upper(char c) {
    return c >= 'A' && c <= 'Z';
}

bool is_digit(char c) {
    return c >= '0' && c <= '9';
}

bool is_ident_char(char c) {
    return is_lower(c) || is_upper(c) || is_digit(c) || c == '_';
}

bool is_identifier(const std::string& text) {
    if (text.empty()) {
        return false;
    }
    if (is_digit(text.front())) {
        return false;
    }
    return std::all_of(text.begin(), text.end(), is_ident_char);
}

std::string to_lower(std::string text) {
    std::transform(text.begin(), text.end(), text.begin(), [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    });
    return text;
}

std::string trim(const std::string& text) {
    const auto first = text.find_first_not_of(" \t\r\n");
    if (first == std::string::npos) {
        return {};
    }
    const auto last = text.find_last_not_of(" \t\r\n");
    return text.substr(first, last - first + 1);
}

// Strip any mix of leading/trailing angle brackets and ampersands, which is how
// x64dbg wraps import references: <&kernel32.VirtualAllocEx>
std::string strip_wrappers(std::string text) {
    text = trim(text);
    while (!text.empty() && (text.front() == '<' || text.front() == '&')) {
        text.erase(text.begin());
    }
    while (!text.empty() && (text.back() == '>' || text.back() == '&')) {
        text.pop_back();
    }
    return trim(text);
}

bool starts_with(const std::string& text, const std::string& prefix) {
    return text.size() >= prefix.size() &&
           text.compare(0, prefix.size(), prefix) == 0;
}

bool ends_with(const std::string& text, const std::string& suffix) {
    return text.size() >= suffix.size() &&
           text.compare(text.size() - suffix.size(), suffix.size(), suffix) == 0;
}

// Native API prefixes. These take UNICODE_STRING and have no ANSI or Unicode
// spellings, so expanding A/W siblings for them produces keys that never hit.
constexpr std::array<const char*, 7> kNativePrefixes = {
    "Nt", "Zw", "Rtl", "Ldr", "Ke", "Csr", "Dbg"};

// Modules that forward to, or are implementation details of, another module.
// Advisory only -- module never filters a lookup, because apiset and forwarder
// chains make a strict DLL match wrong more often than right.
struct ModuleAlias {
    const char* from;
    const char* to;
};

constexpr std::array<ModuleAlias, 4> kModuleAliases = {{
    {"kernelbase", "kernel32"},
    {"ntoskrnl", "ntdll"},
    {"wow64win", "win32u"},
    {"sechost", "advapi32"},
}};

void push_unique(std::vector<std::string>& keys, const std::string& key) {
    if (key.empty()) {
        return;
    }
    if (std::find(keys.begin(), keys.end(), key) == keys.end()) {
        keys.push_back(key);
    }
}

// Ordered exact-spelling candidates, most specific first.
//
// Matched against api.name, so they keep their original case; the case-folded
// fallback is Symbol::canonical_key and is tried only after these all miss.
//
// The exact spelling comes first so CreateProcessA lands on the ANSI entry
// directly rather than arriving through the base name. Charset siblings expand
// in both directions: W->A reaches MalAPI's 122 ANSI-only entries from the
// spelling seen at runtime, and bare->A/W lets a capa rule written as
// `api: CreateProcess` match a row stored as CreateProcessA.
std::vector<std::string> build_lookup_keys(const std::string& function) {
    std::vector<std::string> keys;
    const std::string base = fold_charset_suffix(function);

    push_unique(keys, function);
    push_unique(keys, fold_zw_to_nt(function));
    push_unique(keys, base);

    if (!is_native_api(function)) {
        push_unique(keys, base + "A");
        push_unique(keys, base + "W");
    }
    return keys;
}

}  // namespace

std::string strip_decoration(const std::string& raw) {
    std::string name = trim(raw);

    static const std::array<const char*, 4> kPrefixes = {
        "__imp__", "__imp_", "_imp__", "_imp_"};
    for (const char* prefix : kPrefixes) {
        if (starts_with(name, prefix)) {
            name.erase(0, std::string(prefix).size());
            break;
        }
    }

    // 32-bit stdcall decoration: _VirtualAllocEx@20
    if (name.size() > 2 && name.front() == '_') {
        const auto at = name.rfind('@');
        if (at != std::string::npos && at + 1 < name.size()) {
            const bool all_digits = std::all_of(
                name.begin() + static_cast<long>(at) + 1, name.end(), is_digit);
            const std::string inner = name.substr(1, at - 1);
            if (all_digits && is_identifier(inner)) {
                name = inner;
            }
        }
    }
    return name;
}

std::string fold_charset_suffix(const std::string& name) {
    if (name.size() < 2) {
        return name;
    }
    const char last = name.back();
    if (last != 'A' && last != 'W') {
        return name;
    }

    const char preceding = name[name.size() - 2];
    if (is_lower(preceding) || is_digit(preceding)) {
        return name.substr(0, name.size() - 1);
    }
    // DnsQuery_A -> DnsQuery, so both spellings share one canonical key.
    if (preceding == '_' && name.size() > 2) {
        return name.substr(0, name.size() - 2);
    }
    return name;
}

std::string fold_zw_to_nt(const std::string& name) {
    if (name.size() > 2 && name[0] == 'Z' && name[1] == 'w' && is_upper(name[2])) {
        return "Nt" + name.substr(2);
    }
    return name;
}

std::string canonical(const std::string& name) {
    // Case-folded last, and deliberately. MalAPI title-cases the Winsock family
    // -- Socket, Accept, Recv, Listen, Closesocket -- while the real exports and
    // Microsoft's pages are lowercase, and it writes GetKeynameTextA where
    // Microsoft has GetKeyNameTextA. A case-sensitive key leaves the intent
    // layer unreachable for all of them. Folding runs first because the A/W and
    // Zw rules need the original case to tell a charset suffix from an acronym.
    return to_lower(fold_zw_to_nt(fold_charset_suffix(strip_decoration(name))));
}

bool is_native_api(const std::string& function) {
    for (const char* prefix : kNativePrefixes) {
        const std::string p(prefix);
        if (function.size() > p.size() && starts_with(function, p) &&
            is_upper(function[p.size()])) {
            return true;
        }
    }
    return false;
}

void normalize_module(const std::string& raw, std::string& out, bool& is_apiset) {
    out.clear();
    is_apiset = false;

    std::string module = to_lower(strip_wrappers(raw));
    if (module.empty()) {
        return;
    }

    static const std::array<const char*, 5> kSuffixes = {
        ".dll", ".exe", ".sys", ".drv", ".lib"};
    for (const char* suffix : kSuffixes) {
        if (ends_with(module, suffix)) {
            module.erase(module.size() - std::string(suffix).size());
            break;
        }
    }

    if (starts_with(module, "api-ms-win-") || starts_with(module, "ext-ms-win-")) {
        is_apiset = true;
        return;
    }
    if (module.empty()) {
        return;
    }

    for (const ModuleAlias& alias : kModuleAliases) {
        if (module == alias.from) {
            out = alias.to;
            return;
        }
    }
    out = module;
}

Symbol parse_symbol(const std::string& raw) {
    Symbol symbol;
    symbol.raw = raw;

    std::string text = trim(raw);
    if (text.empty()) {
        return symbol;
    }

    // Pull the symbol out of a memory operand: ds:[<&CreateProcessW>]
    const auto open = text.find('[');
    if (open != std::string::npos) {
        const auto close = text.find(']', open + 1);
        if (close != std::string::npos) {
            text = text.substr(open + 1, close - open - 1);
        }
    }

    text = strip_wrappers(text);

    // x64dbg prefixes IAT thunks with the branch mnemonic: JMP.&GetProcAddress
    const std::string lowered = to_lower(text);
    for (const char* mnemonic : {"jmp.", "call."}) {
        if (starts_with(lowered, mnemonic)) {
            text = text.substr(std::string(mnemonic).size());
            break;
        }
    }

    text = strip_wrappers(text);
    if (text.empty()) {
        return symbol;
    }

    // Split module from function on the last dot, but only when the tail is not
    // a file extension -- "kernel32.dll" alone must not yield function "dll".
    // Module names may themselves contain dots; apiset stubs do.
    std::string module_part;
    std::string function_part = text;
    const auto dot = text.rfind('.');
    if (dot != std::string::npos) {
        const std::string tail = to_lower(text.substr(dot + 1));
        if (tail == "dll" || tail == "exe" || tail == "sys" || tail == "drv") {
            module_part = text;
            function_part.clear();
        } else if (dot > 0) {
            module_part = text.substr(0, dot);
            function_part = text.substr(dot + 1);
        }
    }

    normalize_module(module_part, symbol.module, symbol.is_apiset);

    if (function_part.empty()) {
        return symbol;
    }

    if (function_part.front() == '#') {
        const std::string digits = function_part.substr(1);
        if (!digits.empty() && std::all_of(digits.begin(), digits.end(), is_digit)) {
            symbol.ordinal = std::stoi(digits);
        }
        return symbol;
    }

    function_part = strip_decoration(function_part);
    if (!is_identifier(function_part)) {
        return symbol;
    }

    symbol.function = function_part;
    symbol.lookup_keys = build_lookup_keys(function_part);
    symbol.canonical_key = canonical(function_part);
    return symbol;
}

}  // namespace malapi

File: etl/normalize.py
"""Windows API name normalization.

Single source of truth for turning whatever x64dbg puts in front of the analyst --
an export label, an IAT thunk, a stdcall-decorated symbol -- into an ordered list
of keys to look up in the index.

Two facts about the source data drive everything here:

  * malapi.json lists 122 APIs under their ``A`` spelling and *none* under ``W``.
    At runtime you overwhelmingly see ``CreateProcessW``. Without A/W folding the
    panel is blank for the common case.
  * ntdll exports ``NtX`` and ``ZwX`` at the same address, and MalAPI only ever
    uses the ``Nt`` spelling.

``src/normalize.cpp`` mirrors this file. Both are asserted against
``tests/normalize_cases.json`` -- when behaviour needs to change, change the
fixture table first, then both implementations.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

# Apiset stub names (api-ms-win-core-processthreads-l1-1-1, ext-ms-win-...) never
# name the DLL that actually implements the call, so they are recorded and then
# ignored for lookup purposes.
_APISET_RE = re.compile(r"^(?:api|ext)-ms-win-", re.IGNORECASE)

# 32-bit stdcall decoration: _VirtualAllocEx@20
_STDCALL_RE = re.compile(r"^_(?P<name>[A-Za-z_][A-Za-z0-9_]*)@\d+$")

# Ordinal-only exports: #123 or kernel32.#123
_ORDINAL_RE = re.compile(r"^#(?P<ordinal>\d+)$")

# A bare identifier. Anything that fails this is not a symbol we can look up.
_IDENT_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")

# Modules that forward to, or are implementation details of, another module.
# Used only to make the panel say something sensible about provenance -- module
# is advisory metadata, never a lookup filter, because apiset and forwarder
# chains make any strict DLL match wrong more often than right.
_MODULE_ALIASES = {
    "kernelbase": "kernel32",
    "ntoskrnl": "ntdll",
    "wow64win": "win32u",
    "sechost": "advapi32",
}

# Native API prefixes. These functions take UNICODE_STRING and have no ANSI or
# Unicode spellings, so expanding A/W siblings for them yields keys that can
# never hit (NtOpenProcessA does not exist).
_NATIVE_PREFIXES = ("Nt", "Zw", "Rtl", "Ldr", "Ke", "Csr", "Dbg")

# Function names ending in an uppercase A or W that are NOT charset variants.
# Empty today; kept as the documented escape hatch so a future false fold has an
# obvious home rather than becoming a special case inside fold_charset_suffix.
_NOT_CHARSET_VARIANTS: frozenset[str] = frozenset()


@dataclass
class Symbol:
    """A parsed symbol reference, split into its addressable parts."""

    raw: str
    function: str = ""
    module: str | None = None
    ordinal: int | None = None
    is_apiset: bool = False
    #: Case-preserving exact spellings to try against ``api.name``, most
    #: specific first.
    lookup_keys: list[str] = field(default_factory=list)
    #: Single case-folded key to try against ``api.name_norm`` once every exact
    #: spelling has missed. All of ``lookup_keys`` reduce to this one value.
    canonical_key: str = ""

    @property
    def resolvable(self) -> bool:
        return bool(self.lookup_keys)


def strip_decoration(name: str) -> str:
    """Remove import-thunk and stdcall decoration from a symbol name.

    ``__imp_CreateProcessW`` -> ``CreateProcessW``
    ``_VirtualAllocEx@20``   -> ``VirtualAllocEx``
    """
    name = name.strip()
    for prefix in ("__imp__", "__imp_", "_imp__", "_imp_"):
        if name.startswith(prefix):
            name = name[len(prefix) :]
            break

    stdcall = _STDCALL_RE.match(name)
    if stdcall:
        name = stdcall.group("name")

    return name


def fold_charset_suffix(name: str) -> str:
    """Fold an ANSI/Unicode variant onto its base name.

    ``CreateProcessW`` and ``CreateProcessA`` both fold to ``CreateProcess``;
    ``CopyFileExW`` folds to ``CopyFileEx``. The ``Ex`` suffix is deliberately
    preserved -- ``VirtualAlloc`` and ``VirtualAllocEx`` are different functions
    and must never collapse together.

    The trailing letter only counts as a charset marker when the character before
    it is lowercase, a digit, or an underscore. That keeps all-caps acronyms
    intact -- a hypothetical ``RSA`` ends in ``A`` but ``S`` is uppercase, so it
    is left alone -- while still catching dnsapi's underscore-separated
    ``DnsQuery_A`` / ``DnsQuery_W`` pair, which is otherwise unreachable.
    """
    if name in _NOT_CHARSET_VARIANTS:
        return name
    if len(name) < 2 or name[-1] not in ("A", "W"):
        return name

    preceding = name[-2]
    if preceding.islower() or preceding.isdigit():
        return name[:-1]
    if preceding == "_" and len(name) > 2:
        # DnsQuery_A -> DnsQuery, so both spellings share one canonical key.
        return name[:-2]
    return name


def fold_zw_to_nt(name: str) -> str:
    """Fold the ``Zw`` spelling of a native API onto its ``Nt`` twin.

    ntdll exports both at the same address and MalAPI only ever uses ``Nt``, so
    ``ZwOpenProcess`` has to reach the ``NtOpenProcess`` entry. Guarded on a
    following uppercase letter so an unrelated name starting with "Zw" is safe.
    """
    if len(name) > 2 and name.startswith("Zw") and name[2].isupper():
        return "Nt" + name[2:]
    return name


def canonical(name: str) -> str:
    """Reduce a function name to the key stored in ``api.name_norm``.

    Applied identically at index-build time and at lookup time, which is what
    makes the two sides meet.

    Case-folded last, and deliberately. MalAPI title-cases the Winsock family --
    ``Socket``, ``Accept``, ``Recv``, ``Listen``, ``Closesocket`` -- while the
    real exports and Microsoft's own pages are lowercase, and it writes
    ``GetKeynameTextA`` where Microsoft has ``GetKeyNameTextA``. A case-sensitive
    key leaves the malicious-intent layer unreachable for every one of them.

    Folding happens before lowercasing because the A/W and Zw rules need the
    original case to tell a charset suffix from an acronym.
    """
    return fold_zw_to_nt(fold_charset_suffix(strip_decoration(name))).lower()


def normalize_module(module: str | None) -> tuple[str | None, bool]:
    """Normalize a module name, reporting whether it was an apiset stub.

    Returns ``(module_or_None, is_apiset)``. Apiset stubs resolve to ``None``
    because the stub name tells you nothing about the implementing DLL.
    """
    if not module:
        return None, False

    module = module.strip().strip("<>&").lower()
    for suffix in (".dll", ".exe", ".sys", ".drv", ".lib"):
        if module.endswith(suffix):
            module = module[: -len(suffix)]

    if _APISET_RE.match(module):
        return None, True
    if not module:
        return None, False

    return _MODULE_ALIASES.get(module, module), False


def parse_symbol(raw: str) -> Symbol:
    """Parse anything x64dbg might hand us into a :class:`Symbol`.

    Handles the shapes that actually turn up in the disassembly view::

        VirtualAllocEx
        kernel32.VirtualAllocEx
        <kernel32.VirtualAllocEx>
        JMP.&GetProcAddress
        qword ptr ds:[<&CreateProcessW>]
        __imp_CreateProcessW
        _VirtualAllocEx@20
        kernel32.#123
    """
    sym = Symbol(raw=raw)
    text = (raw or "").strip()
    if not text:
        return sym

    # Pull the symbol out of a memory operand: ds:[<&CreateProcessW>] and friends.
    bracketed = re.search(r"\[([^\[\]]+)\]", text)
    if bracketed:
        text = bracketed.group(1)

    text = text.strip().strip("<>").lstrip("&").strip()

    # x64dbg prefixes IAT thunks with the branch mnemonic: JMP.&GetProcAddress
    for mnemonic in ("jmp.", "call."):
        if text.lower().startswith(mnemonic):
            text = text[len(mnemonic) :].lstrip("&")
            break

    text = text.strip().strip("<>").lstrip("&").strip()
    if not text:
        return sym

    # Split module from function on the last dot, but only when the tail is not
    # itself a file extension -- "kernel32.dll" alone must not become function
    # "dll". Module names themselves may contain dots (apiset stubs do).
    module_part: str | None = None
    function_part = text
    if "." in text:
        head, _, tail = text.rpartition(".")
        if tail.lower() in ("dll", "exe", "sys", "drv"):
            module_part, function_part = text, ""
        elif head:
            module_part, function_part = head, tail

    sym.module, sym.is_apiset = normalize_module(module_part)

    if not function_part:
        return sym

    ordinal = _ORDINAL_RE.match(function_part)
    if ordinal:
        sym.ordinal = int(ordinal.group("ordinal"))
        return sym

    function_part = strip_decoration(function_part)
    if not _IDENT_RE.match(function_part):
        return sym

    sym.function = function_part
    sym.lookup_keys = _build_lookup_keys(function_part)
    sym.canonical_key = canonical(function_part)
    return sym


def is_native_api(function: str) -> bool:
    """True for ntdll-style natives, which have no ANSI/Unicode variants."""
    return any(
        function.startswith(p) and len(function) > len(p) and function[len(p)].isupper()
        for p in _NATIVE_PREFIXES
    )


def _build_lookup_keys(function: str) -> list[str]:
    """Ordered exact-spelling candidates, most specific first.

    These are matched against ``api.name``, so they keep their original case; the
    case-folded fallback lives in :attr:`Symbol.canonical_key` and is tried only
    after all of these miss.

    The exact spelling comes first, so ``CreateProcessA`` lands on the ANSI entry
    directly instead of arriving there via the base name. Charset siblings expand
    in both directions: W->A reaches MalAPI's 122 ANSI-only entries from the
    spelling seen at runtime, and bare->A/W lets a capa rule written as
    ``api: CreateProcess`` match a row stored as ``CreateProcessA``.
    """
    keys: list[str] = []

    def push(key: str) -> None:
        if key and key not in keys:
            keys.append(key)

    base = fold_charset_suffix(function)

    push(function)
    push(fold_zw_to_nt(function))
    push(base)

    if not is_native_api(function):
        push(base + "A")
        push(base + "W")

    return keys


def lookup_keys(raw: str) -> list[str]:
    """Convenience wrapper: parse ``raw`` and return just its lookup keys."""
    return parse_symbol(raw).lookup_keys
